core/src/hir/types.rs

```diff
@@ -0,0 +1,88 @@
+//! Types that can be exposed in Diplomat APIs.
+
+use super::{
+    EnumId, OpaqueId, PrimitiveType, StructId, StructIdKind, TypeContext, TypeLifetime,
+    TypeLifetimes,
+};
+
+/// Type that may be used as an output.
+pub enum TypeKind {
+    OutType(OutType),
+    Type(Type),
+}
+
+/// Type that can only be used as an output.
+pub enum OutType {
+    Primitive(PrimitiveType),
+    Opaque(TypeLifetimes, Optionality, Ownership, OpaqueId),
+    Struct(TypeLifetimes, StructIdKind),
+    Enum(EnumId),
+    Slice(Slice),
+}
+
+/// Type that may be used as input or output.
+pub enum Type {
+    Primitive(PrimitiveType),
+    Opaque(TypeLifetimes, Optionality, TypeLifetime, OpaqueId),
```

Comment:
nit: we should probably call these OpaqueRef everywhere

and perhaps we have an OpaqueRef type somewhere that encapsulates these? Enums with four fields are tricky

Reply:
I named the struct that goes in the enum to be `OpaqueRef` and `ReturnedOpaque`, but I left the variant name as just `Opaque` because I thought it looked bad lol. Thoughts?